### DISCUSSION TITLE
useSigner cache in localstorage and returned as a deserialized object

### DISCUSSION THREAD

**Discussion by gravityblast** [wevm/wagmi#907] [Q&A]
Hey everyone, 

I'm struggling to fix a problem I have using `useSigner`, but after a day on that I hope someone can help me :)

This happens in a big repo and I couldn't reproduce it in a small example app unfortunately.

In my react component I'm getting the signer with  `const { data: signer } = useSigner();` and then I pass it to `ethers` as a `signerOrProvider`.

Just in a specific case, it happens that `signer` is a plain object instead of being an instance of `JsonRpcObject`, so `ethers` fails because it doesn't have the `getAddress` method.

After debugging the code I actually realized that sometimes the query is cached under the `wagmi.cache` key in localstorage.

```
queryHash: "[{\"entity\":\"signer\"}]" 
queryKey: [{entity: "signer"}] 
state: {data: {_isSigner: true,…}, dataUpdateCount: 156, dataUpdatedAt: 1662057694466, error: null,…}
```

For this reason returned object is a plain deserialized object instead of being an instance of `JsonRpcObject`.

Looking at the code it looks like that query should never be cached, but in my case it happens if I:

* connect with metamask
* disconnect
* connect with rainbow (using `rainbowkit`)
* disconnect
* connect with metamask

At this point I see the query cached in local storage and the error.

Debugging the code I see that `useSigner` is called multiple times with different results:

1 - `signer` is a plain deserialized object
2 - `signer` is `null`
3 - `signer` is `null`
4 - `signer` is finally a `JsonRpcSigner`

As a quick fix I'm now checking that the object has the `getAddress` method, but I guess the problem is somewhere else.

Did this happen to anyone else?

❤️ 🌈

**Comment by jxom** [comment]
Yeah... The `useSigner` hook isn't meant to be cached. There is an open PR to fix that here: https://github.com/wagmi-dev/wagmi/pull/904

**Reply by wjpg** [reply]
@jxom any idea when #904 will be merged & released? 🙏

**Comment by gravityblast** [comment]
hey thank you @jxom , that's great!
do you know when does that happen? cause without the steps I descrbed it's actually never cached.

**Reply by jxom** [reply]
tbh, it shouldn't at all since we force the signer query cache time to zero. Might be a bug in React Query somewhere. 😅
I'll need to do some digging on that side when I get the time.
